// @spicygolf/ghin/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AuthenticationError: () => AuthenticationError,
  CacheError: () => CacheError,
  ConfigurationError: () => ConfigurationError,
  ErrorCodes: () => ErrorCodes,
  GhinClient: () => GhinClient,
  GhinError: () => GhinError,
  NetworkError: () => NetworkError,
  RateLimitError: () => RateLimitError,
  ValidationError: () => ValidationError,
  boolean: () => boolean,
  date: () => date,
  emptyStringToNull: () => emptyStringToNull,
  float: () => float,
  gender: () => gender,
  handicap: () => handicap,
  monthDay: () => monthDay,
  number: () => number,
  schemaCacheClient: () => schemaCacheClient,
  schemaClientConfig: () => schemaClientConfig,
  schemaCourseCountriesResponse: () => schemaCourseCountriesResponse,
  schemaCourseCountry: () => schemaCourseCountry,
  schemaCourseCountryCode: () => schemaCourseCountryCode,
  schemaCourseDetailsRequest: () => schemaCourseDetailsRequest,
  schemaCourseDetailsResponse: () => schemaCourseDetailsResponse,
  schemaCourseHandicapsRequest: () => schemaCourseHandicapsRequest,
  schemaCoursePercentPlayerHandicap: () => schemaCoursePercentPlayerHandicap,
  schemaCoursePlayerHandicapsResponse: () => schemaCoursePlayerHandicapsResponse,
  schemaCourseSearchRequest: () => schemaCourseSearchRequest,
  schemaCourseSearchResponse: () => schemaCourseSearchResponse,
  schemaFacility: () => schemaFacility,
  schemaFacilityCourse: () => schemaFacilityCourse,
  schemaFacilitySearchRequest: () => schemaFacilitySearchRequest,
  schemaFacilitySearchResponse: () => schemaFacilitySearchResponse,
  schemaGolfer: () => schemaGolfer,
  schemaGolferCourseHandicapRequest: () => schemaGolferCourseHandicapRequest,
  schemaGolferHandicapResponse: () => schemaGolferHandicapResponse,
  schemaGolfersGlobalSearchRequest: () => schemaGolfersGlobalSearchRequest,
  schemaGolfersSearchRequest: () => schemaGolfersSearchRequest,
  schemaGolfersSearchResponse: () => schemaGolfersSearchResponse,
  schemaHoleDetail: () => schemaHoleDetail,
  schemaPlayerCourseHandicap: () => schemaPlayerCourseHandicap,
  schemaScoresRequest: () => schemaScoresRequest,
  schemaScoresResponse: () => schemaScoresResponse,
  schemaScoringAdjustment: () => schemaScoringAdjustment,
  schemaStatistics: () => schemaStatistics,
  schemaTeeSetRatingRequest: () => schemaTeeSetRatingRequest,
  schemaTeeSetRatingResponse: () => schemaTeeSetRatingResponse,
  schemaTeeSetSide: () => schemaTeeSetSide,
  shortDate: () => shortDate,
  string: () => string
});
module.exports = __toCommonJS(index_exports);

// src/client/ghin/index.ts
var import_zod27 = require("zod");

// src/errors/index.ts
var GhinError = class extends Error {
  code;
  statusCode;
  cause;
  constructor(message, code, statusCode, cause) {
    super(message);
    this.name = "GhinError";
    this.code = code;
    this.statusCode = statusCode;
    this.cause = cause;
  }
};
var AuthenticationError = class extends GhinError {
  constructor(message, statusCode, cause) {
    super(message, "AUTHENTICATION_ERROR", statusCode, cause);
    this.name = "AuthenticationError";
  }
};
var NetworkError = class extends GhinError {
  constructor(message, statusCode, cause) {
    super(message, "NETWORK_ERROR", statusCode, cause);
    this.name = "NetworkError";
  }
};
var ValidationError = class extends GhinError {
  field;
  response;
  constructor(message, field, cause, response) {
    super(message, "VALIDATION_ERROR", void 0, cause);
    this.name = "ValidationError";
    this.field = field;
    this.response = JSON.stringify(response);
  }
};
var RateLimitError = class extends GhinError {
  retryAfter;
  constructor(message, retryAfter, cause) {
    super(message, "RATE_LIMIT_ERROR", 429, cause);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
};
var ConfigurationError = class extends GhinError {
  constructor(message, cause) {
    super(message, "CONFIGURATION_ERROR", void 0, cause);
    this.name = "ConfigurationError";
  }
};
var CacheError = class extends GhinError {
  constructor(message, cause) {
    super(message, "CACHE_ERROR", void 0, cause);
    this.name = "CacheError";
  }
};
var ErrorCodes = {
  AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR",
  NETWORK_ERROR: "NETWORK_ERROR",
  VALIDATION_ERROR: "VALIDATION_ERROR",
  RATE_LIMIT_ERROR: "RATE_LIMIT_ERROR",
  CONFIGURATION_ERROR: "CONFIGURATION_ERROR",
  CACHE_ERROR: "CACHE_ERROR"
};

// src/models/cache-client.ts
var import_zod = require("zod");
var schemaStringOrUndefined = import_zod.z.union([import_zod.z.string(), import_zod.z.undefined()]);
var schemaPromiseOrNonPromiseStringOrUndefined = import_zod.z.union([
  import_zod.z.promise(schemaStringOrUndefined),
  schemaStringOrUndefined
]);
var schemaPromiseOrNonPromiseVoid = import_zod.z.union([import_zod.z.promise(import_zod.z.void()), import_zod.z.void()]);
var schemaCacheClient = import_zod.z.object({
  read: import_zod.z.function().args().returns(schemaPromiseOrNonPromiseStringOrUndefined),
  write: import_zod.z.function().args(import_zod.z.string()).returns(schemaPromiseOrNonPromiseVoid)
});

// src/models/client-config.ts
var import_zod3 = require("zod");

// src/models/validation.ts
var import_date_fns = require("date-fns");
var import_zod2 = require("zod");
var boolean = import_zod2.z.union([import_zod2.z.boolean(), import_zod2.z.literal("true"), import_zod2.z.literal("false"), import_zod2.z.null()]).transform((value) => value === true || value === "true");
var date = import_zod2.z.union([import_zod2.z.date(), import_zod2.z.string(), import_zod2.z.null(), import_zod2.z.undefined()]).refine(
  (value) => {
    if (value === null || value === void 0 || value === "") {
      return true;
    }
    if (value instanceof Date) {
      return (0, import_date_fns.isValid)(value);
    }
    if (typeof value === "string") {
      let parsed = (0, import_date_fns.parseISO)(value);
      if ((0, import_date_fns.isValid)(parsed)) {
        return true;
      }
      const formats = [
        "yyyy-MM-dd",
        "yyyy/MM/dd",
        "MM/dd/yyyy",
        "dd/MM/yyyy",
        "MMMM dd, yyyy",
        // September 15, 2022
        "MMM dd, yyyy",
        // Sep 15, 2022
        "MMM dd yyyy",
        // Sep 15 2022
        "dd MMM yyyy"
        // 15 Sep 2022
      ];
      return formats.some((format) => {
        try {
          parsed = (0, import_date_fns.parse)(value, format, /* @__PURE__ */ new Date());
          return (0, import_date_fns.isValid)(parsed);
        } catch {
          return false;
        }
      });
    }
    return false;
  },
  {
    message: "Invalid date"
  }
).transform((value) => {
  if (value === null || value === void 0 || value === "") {
    return void 0;
  }
  if (value instanceof Date) {
    return value;
  }
  if (typeof value === "string") {
    let parsed = (0, import_date_fns.parseISO)(value);
    if ((0, import_date_fns.isValid)(parsed)) {
      return parsed;
    }
    const formats = [
      "yyyy-MM-dd",
      "yyyy/MM/dd",
      "MM/dd/yyyy",
      "dd/MM/yyyy",
      "MMMM dd, yyyy",
      // September 15, 2022
      "MMM dd, yyyy",
      // Sep 15, 2022
      "MMM dd yyyy",
      // Sep 15 2022
      "dd MMM yyyy"
      // 15 Sep 2022
    ];
    for (const format of formats) {
      try {
        parsed = (0, import_date_fns.parse)(value, format, /* @__PURE__ */ new Date());
        if ((0, import_date_fns.isValid)(parsed)) {
          return parsed;
        }
      } catch {
      }
    }
  }
  return void 0;
});
var emptyString = import_zod2.z.string().trim();
var emptyStringToNull = emptyString.nullable().transform((value) => value || null);
var float = import_zod2.z.coerce.number();
var gender = import_zod2.z.enum(["M", "F"]);
var handicap = import_zod2.z.union([float, import_zod2.z.string(), import_zod2.z.null()]).refine((value) => {
  if (typeof value === "number") {
    return true;
  }
  if (typeof value === "string" && (value === "NH" || value === "-")) {
    return true;
  }
  return false;
}).transform((value) => {
  if (value === "NH" || value === "-") {
    return null;
  }
  return value;
});
var number = float.int();
var string = emptyString.min(1);
var monthDay = string.or(emptyString).transform((value) => {
  if (!value) {
    return null;
  }
  const [month, day] = value.split("/");
  return `${month?.toString().padStart(2, "0")}-${day?.toString().padStart(2, "0")}`;
});
var shortDate = import_zod2.z.union([import_zod2.z.date(), import_zod2.z.string(), import_zod2.z.null()]).refine((value) => value ? !Number.isNaN(Date.parse(value.toString())) : true, {
  message: "Invalid date"
}).transform((value) => {
  if (typeof value !== "string") {
    return value;
  }
  const [year, month, day] = value.split("-");
  return /* @__PURE__ */ new Date(`${year}-${month}-${day}T00:00Z`);
});

// src/models/client-config.ts
var schemaClientConfig = import_zod3.z.object({
  apiAccess: boolean.optional(),
  apiVersion: string.optional(),
  baseUrl: string.optional(),
  cache: schemaCacheClient.optional(),
  password: string,
  username: string
});

// src/client/in-memory-cache-client/index.ts
var InMemoryCacheClient = class {
  value = void 0;
  async read() {
    return this.value;
  }
  async write(value) {
    this.value = value;
  }
};

// src/client/request-client/index.ts
var import_async_mutex = require("async-mutex");
var import_jwt_decode = require("jwt-decode");
var import_neverthrow2 = require("neverthrow");

// src/utils/retry.ts
var import_neverthrow = require("neverthrow");
var DEFAULT_RETRY_CONFIG = {
  maxAttempts: 3,
  baseDelay: 1e3,
  // 1 second
  maxDelay: 1e4,
  // 10 seconds
  backoffMultiplier: 2,
  retryableStatusCodes: [408, 429, 500, 502, 503, 504]
};
function isRetryableError(error) {
  if (error instanceof NetworkError) {
    return error.statusCode ? DEFAULT_RETRY_CONFIG.retryableStatusCodes.includes(error.statusCode) : true;
  }
  if (error instanceof RateLimitError) {
    return true;
  }
  return false;
}
function calculateDelay(attempt, config) {
  const delay = config.baseDelay * config.backoffMultiplier ** (attempt - 1);
  return Math.min(delay, config.maxDelay);
}
async function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
async function withRetry(operation, config = {}) {
  const finalConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  for (let attempt = 1; attempt <= finalConfig.maxAttempts; attempt++) {
    const result = await operation();
    if (result.isOk()) {
      return result;
    }
    const error = result.error;
    if (attempt === finalConfig.maxAttempts) {
      return result;
    }
    if (!isRetryableError(error)) {
      return result;
    }
    const delay = calculateDelay(attempt, finalConfig);
    const jitter = Math.random() * 0.1 * delay;
    const totalDelay = delay + jitter;
    await sleep(totalDelay);
  }
  return (0, import_neverthrow.err)(new Error("Retry exhausted"));
}

// src/client/request-client/models/access-token.ts
var import_zod4 = require("zod");
var schemaAccessToken = import_zod4.z.object({
  expiresIn: string,
  token: string
}).transform(({ expiresIn: tmpExpiresIn, ...values }) => {
  const now = /* @__PURE__ */ new Date();
  const secondsUntilExpiry = Number(tmpExpiresIn.replace(/[^0-9]/g, ""));
  const millisecondsUntilExpiry = secondsUntilExpiry * 1e3;
  const expiresIn = new Date(now.getTime() + millisecondsUntilExpiry);
  return {
    ...values,
    expiresIn
  };
});

// src/client/request-client/models/session.ts
var import_zod5 = require("zod");
var schemaSession = import_zod5.z.object({
  appId: string,
  authVersion: string,
  fid: string,
  sdkVersion: string
});
var schemaSessionResponse = import_zod5.z.object({
  authToken: schemaAccessToken
});

// src/client/request-client/models/login.ts
var import_zod6 = require("zod");
var schemaLoginAPIRequest = import_zod6.z.object({
  user: import_zod6.z.object({
    email: string,
    password: string,
    remember_me: boolean
  })
});
var schemaLoginResponse = import_zod6.z.object({
  golfer_user: import_zod6.z.object({
    golfer_user_token: string
  })
});
var schemaLoginAPIResponse = import_zod6.z.object({
  user: import_zod6.z.object({
    id: string,
    email: emptyStringToNull.optional(),
    prefix: emptyStringToNull.optional(),
    first_name: emptyStringToNull.optional(),
    middle_name: emptyStringToNull.optional(),
    last_name: emptyStringToNull.optional(),
    suffix: emptyStringToNull.optional(),
    phone: emptyStringToNull.optional(),
    last_sign_in_at: emptyStringToNull.optional()
  }),
  token: string
});

// src/client/request-client/index.ts
var FIREBASE_SESSION_URL = new URL(
  "https://firebaseinstallations.googleapis.com/v1/projects/ghin-mobile-app/installations"
);
var GOOGLE_API_KEY = "AIzaSyBxgTOAWxiud0HuaE5tN-5NTlzFnrtyz-I";
var DEFAULT_USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36";
var CLIENT_SOURCE = "GHINcom";
var SESSION_DEFAULTS = {
  appId: "1:884417644529:web:47fb315bc6c70242f72650",
  authVersion: "FIS_v2",
  fid: "fg6JfS0U01YmrelthLX9Iz",
  sdkVersion: "w:0.5.7"
};
var FETCH_HEADER_DEFAULTS = {
  "Content-Type": "application/json",
  "User-Agent": DEFAULT_USER_AGENT
};
var apiPathnames = {
  course_countries: "/get_countries_and_states.json",
  course_details: "/crsCourseMethods.asmx/GetCourseDetails.json",
  course_handicaps: "/playing_handicaps.json",
  course_search: "/crsCourseMethods.asmx/SearchCourses.json",
  facility_search: "/facilities/search.json",
  golfer: "/search_golfer.json",
  golfers_search: "/golfers/search.json",
  golfers_global_search: "/golfers.json",
  login: "/golfer_login.json",
  scores: "/scores.json",
  users_login: "/users/login.json"
};
var toFullApiUrl = (baseUrl, pathname) => new URL(`${baseUrl.pathname}${apiPathnames[pathname]}`, baseUrl);
var makeAuthHeaders = (accessToken) => ({
  Authorization: `Bearer ${accessToken}`
});
var RequestClient = class {
  accessToken;
  baseUrl;
  config;
  lock;
  sessionToken;
  constructor(config) {
    const results = schemaClientConfig.safeParse(config);
    if (!results.success) {
      throw new ConfigurationError(`Invalid RequestClientConfig: ${results.error.message}`);
    }
    if (results.data.apiAccess === void 0) {
      results.data.apiAccess = false;
    }
    if (results.data.apiVersion === void 0) {
      results.data.apiVersion = "v1";
    }
    if (results.data.baseUrl === void 0) {
      results.data.baseUrl = "https://api2.ghin.com/api";
    }
    if (!results.data.cache) {
      results.data.cache = new InMemoryCacheClient();
    }
    this.lock = new import_async_mutex.Mutex();
    this.config = schemaClientConfig.parse(results.data);
    this.baseUrl = new URL(`${this.config.baseUrl}/${this.config.apiVersion}`);
  }
  async _fetch({
    options,
    schema,
    url
  }) {
    try {
      const response = await fetch(url.toString(), options);
      if (!response.ok || response.status >= 400) {
        let body;
        try {
          body = await response.json();
        } catch {
          body = await response.text();
        }
        if (response.status === 401 || response.status === 403) {
          return (0, import_neverthrow2.err)(
            new AuthenticationError(
              `Authentication failed: ${response.status} ${response.statusText}`,
              response.status,
              new Error(JSON.stringify(body))
            )
          );
        }
        if (response.status === 429) {
          const retryAfter = response.headers.get("Retry-After");
          const retryAfterSeconds = retryAfter ? Number.parseInt(retryAfter, 10) : void 0;
          return (0, import_neverthrow2.err)(
            new RateLimitError(
              `Rate limit exceeded: ${response.status} ${response.statusText}`,
              retryAfterSeconds,
              new Error(JSON.stringify(body))
            )
          );
        }
        if (response.status >= 500) {
          return (0, import_neverthrow2.err)(
            new NetworkError(
              `Server error: ${response.status} ${response.statusText}`,
              response.status,
              new Error(JSON.stringify(body))
            )
          );
        }
        return (0, import_neverthrow2.err)(
          new NetworkError(
            `Request failed: ${response.status} ${response.statusText}`,
            response.status,
            new Error(JSON.stringify(body))
          )
        );
      }
      const raw = await response.json();
      const parsed = schema.safeParse(raw);
      if (!parsed.success) {
        return (0, import_neverthrow2.err)(
          new ValidationError(
            `Response validation failed: ${JSON.stringify(parsed.error)}`,
            void 0,
            new Error(`URL: ${url.toString()}`),
            raw
          )
        );
      }
      return (0, import_neverthrow2.ok)(parsed.data);
    } catch (error) {
      if (error instanceof Error) {
        return (0, import_neverthrow2.err)(new NetworkError(`Network request failed: ${error.message}`, void 0, error));
      }
      return (0, import_neverthrow2.err)(new NetworkError(`Unknown network error: ${String(error)}`));
    }
  }
  async refreshSessionToken() {
    const url = new URL(FIREBASE_SESSION_URL);
    const body = JSON.stringify(SESSION_DEFAULTS);
    const result = await this._fetch({
      options: {
        body,
        headers: {
          ...FETCH_HEADER_DEFAULTS,
          "x-goog-api-key": GOOGLE_API_KEY
        },
        method: "POST"
      },
      schema: schemaSessionResponse,
      url
    });
    return result.map((response) => response.authToken);
  }
  isAccessTokenValid(accessToken) {
    if (!accessToken) {
      return false;
    }
    try {
      const decoded = (0, import_jwt_decode.jwtDecode)(accessToken);
      const expirationDate = new Date(decoded.exp * 1e3);
      return expirationDate > /* @__PURE__ */ new Date();
    } catch {
      return false;
    }
  }
  async getAccessToken() {
    const isAccessTokenValid = this.isAccessTokenValid(this.accessToken);
    if (isAccessTokenValid) {
      return (0, import_neverthrow2.ok)(this.accessToken);
    }
    try {
      const cachedAccessToken = await this.config.cache?.read();
      const isCachedTokenValid = this.isAccessTokenValid(cachedAccessToken);
      if (isCachedTokenValid) {
        this.accessToken = cachedAccessToken;
        return (0, import_neverthrow2.ok)(cachedAccessToken);
      }
    } catch (error) {
      return (0, import_neverthrow2.err)(
        new CacheError(
          `Failed to read from cache: ${error instanceof Error ? error.message : String(error)}`,
          error instanceof Error ? error : void 0
        )
      );
    }
    const refreshResult = await this.refreshAccessToken();
    if (refreshResult.isErr()) {
      return refreshResult;
    }
    const accessToken = refreshResult.value;
    this.accessToken = accessToken;
    try {
      await this.config.cache?.write(accessToken);
    } catch (error) {
      return (0, import_neverthrow2.err)(
        new CacheError(
          `Failed to write to cache: ${error instanceof Error ? error.message : String(error)}`,
          error instanceof Error ? error : void 0
        )
      );
    }
    return (0, import_neverthrow2.ok)(accessToken);
  }
  async apiLogin() {
    const url = toFullApiUrl(this.baseUrl, "users_login");
    const body = JSON.stringify({
      user: {
        email: this.config.username,
        password: this.config.password,
        remember_me: true
      }
    });
    const response = await this._fetch({
      options: {
        body,
        headers: FETCH_HEADER_DEFAULTS,
        method: "POST"
      },
      schema: schemaLoginAPIResponse,
      url
    });
    return response.andThen((resp) => {
      if (resp && "token" in resp) {
        return (0, import_neverthrow2.ok)(resp.token);
      }
      return (0, import_neverthrow2.err)(new Error("Login response did not contain a token."));
    });
  }
  async refreshAccessToken() {
    if (this.config.apiAccess) {
      return this.apiLogin();
    }
    const sessionResult = await this.refreshSessionToken();
    if (sessionResult.isErr()) {
      return (0, import_neverthrow2.err)(sessionResult.error);
    }
    this.sessionToken = sessionResult.value;
    const url = toFullApiUrl(this.baseUrl, "login");
    const body = JSON.stringify({
      token: this.sessionToken.token,
      user: {
        email_or_ghin: this.config.username,
        password: this.config.password
      }
    });
    const response = await this._fetch({
      options: {
        body,
        headers: FETCH_HEADER_DEFAULTS,
        method: "POST"
      },
      schema: schemaLoginResponse,
      url
    });
    return response.map((resp) => {
      if ("golfer_user" in resp) {
        return resp.golfer_user.golfer_user_token;
      }
      return resp.token;
    });
  }
  async fetch({
    entity,
    schema,
    options = {}
  }) {
    const accessTokenResult = await this.lock.runExclusive(async () => this.getAccessToken());
    if (accessTokenResult.isErr()) {
      return (0, import_neverthrow2.err)(accessTokenResult.error);
    }
    const accessToken = accessTokenResult.value;
    const url = toFullApiUrl(this.baseUrl, entity);
    const { headers, searchParams, ...requestInitOptions } = options;
    const actualOptions = {
      ...requestInitOptions,
      headers: {
        ...FETCH_HEADER_DEFAULTS,
        source: CLIENT_SOURCE,
        ...makeAuthHeaders(accessToken),
        ...headers
      }
    };
    if (searchParams) {
      url.search = searchParams.toString();
    }
    return withRetry(() => this._fetch({ options: actualOptions, schema, url }));
  }
  async fetchCustomPath({
    path,
    schema,
    options = {}
  }) {
    const accessTokenResult = await this.lock.runExclusive(async () => this.getAccessToken());
    if (accessTokenResult.isErr()) {
      return (0, import_neverthrow2.err)(accessTokenResult.error);
    }
    const accessToken = accessTokenResult.value;
    const url = new URL(`${this.baseUrl.pathname}${path}`, this.baseUrl);
    const { headers, searchParams, ...requestInitOptions } = options;
    const actualOptions = {
      ...requestInitOptions,
      headers: {
        ...FETCH_HEADER_DEFAULTS,
        source: CLIENT_SOURCE,
        ...makeAuthHeaders(accessToken),
        ...headers
      }
    };
    if (searchParams) {
      url.search = searchParams.toString();
    }
    return withRetry(() => this._fetch({ options: actualOptions, schema, url }));
  }
};

// src/client/ghin/models/course/country.ts
var import_zod7 = require("zod");
var FOREIGN_COUNTRY_CODE = "Fo";
var schemaCourseCountryCode = string.refine((value) => {
  if (value === FOREIGN_COUNTRY_CODE) {
    return true;
  }
  return /^[A-Z]{3}$/.test(value);
}, 'Invalid country code. Expected format: ISO 3166-1 alpha-3 code or "Fo" for foreign countries');
var schemaCourseCountryState = import_zod7.z.object({
  code: string,
  //schemaCourseCountryCode,
  course_code: string,
  //schemaCourseCountryCode.nullable(),
  name: string
});
var schemaCourseCountry = import_zod7.z.object({
  code: string,
  crs_code: import_zod7.z.string().trim().transform((value) => value || null).nullable(),
  name: string,
  states: import_zod7.z.array(schemaCourseCountryState)
});

// src/client/ghin/models/course/request.ts
var import_zod8 = require("zod");

// src/client/ghin/models/course/state.ts
var FOREIGN_STATE_CODE = "*F";
var courseStateCodeRegex = /^[A-Z]{1,2}-[A-Z*0-9]{1,4}$/;
var schemaCourseSearchState = string.refine(
  (value) => {
    if (value === FOREIGN_STATE_CODE) {
      return true;
    }
    return courseStateCodeRegex.test(value);
  },
  {
    message: "Invalid state code format. Expected format: US-OH, LC-01, MX-*A, or *F"
  }
);

// src/client/ghin/models/course/request.ts
var schemaCourseSearchRequest = import_zod8.z.object({
  country: schemaCourseCountryCode.optional(),
  facility_id: number.optional(),
  name: string.optional(),
  state: schemaCourseSearchState.optional(),
  updated_at: date.optional()
}).refine(
  ({ country, state, facility_id, updated_at, name }) => {
    switch (true) {
      case Boolean(name):
        return true;
      case Boolean(country && state):
        return true;
      case Boolean(facility_id):
        return true;
      case Boolean(updated_at):
        return true;
      default:
        return false;
    }
  },
  {
    message: "At least one of the following are required: name, country + state, facility_id, or updated_at must be provided"
  }
);
var schemaTeeSetStatus = import_zod8.z.enum(["Active", "Deleted", "All"]);
var schemaCourseDetailsRequest = import_zod8.z.object({
  course_id: number,
  gender: import_zod8.z.enum(["M", "m", "F", "f"]).optional(),
  number_of_holes: import_zod8.z.union([import_zod8.z.literal(9), import_zod8.z.literal(18)]).optional(),
  tee_set_status: schemaTeeSetStatus.optional()
}).transform((data) => ({
  ...data,
  tee_set_status: data.tee_set_status ?? "Active"
}));

// src/client/ghin/models/course/response.ts
var import_zod12 = require("zod");

// src/client/ghin/models/course/course.ts
var import_zod10 = require("zod");

// src/client/ghin/models/course/geolocation.ts
var import_zod9 = require("zod");
var schemaGeoCoordinate = import_zod9.z.preprocess((value) => {
  if (value === void 0 || value === null || value === "") {
    return null;
  }
  if (typeof value === "string") {
    const parsed = Number.parseFloat(value);
    return Number.isNaN(parsed) ? null : parsed;
  }
  if (typeof value === "number") {
    return Number.isNaN(value) ? null : value;
  }
  return null;
}, import_zod9.z.number().nullable());
var schemaGeoAddress = import_zod9.z.preprocess((value) => {
  if (value === void 0 || value === null || value === "") {
    return null;
  }
  return value;
}, import_zod9.z.string().nullable());

// src/client/ghin/models/course/course.ts
var schemaStatus = string.transform((value) => value.toUpperCase()).pipe(import_zod10.z.enum(["ACTIVE", "INACTIVE"]));
var schemaCourse = import_zod10.z.object({
  Address1: string.nullable(),
  Address2: string.nullable(),
  City: string.nullable(),
  Country: string.nullable(),
  CourseID: number,
  CourseName: string,
  CourseStatus: schemaStatus,
  Email: string.email().nullable(),
  EntCountryCode: number.nullable(),
  EntStateCode: number.nullable(),
  FacilityID: number,
  FacilityName: string,
  FacilityStatus: schemaStatus,
  FullName: string,
  GeoLocationLatitude: schemaGeoCoordinate.nullable(),
  GeoLocationLongitude: schemaGeoCoordinate.nullable(),
  LegacyCRPCourseId: number.nullable(),
  Ratings: import_zod10.z.array(import_zod10.z.unknown()),
  State: string.nullable(),
  Telephone: string.nullable(),
  UpdatedOn: shortDate.nullable(),
  Zip: import_zod10.z.string().trim().transform((zip) => zip?.trim() || null).nullable().or(import_zod10.z.array(import_zod10.z.unknown()))
});

// src/client/ghin/models/course/season.ts
var import_zod11 = require("zod");
var schemaSeasonDate = import_zod11.z.preprocess((value) => {
  if (value === void 0 || value === null || value === "") {
    return null;
  }
  return value;
}, import_zod11.z.string().nullable());
var schemaSeasonName = import_zod11.z.preprocess((value) => {
  if (value === void 0 || value === null || value === "") {
    return null;
  }
  return value;
}, import_zod11.z.string().nullable());

// src/client/ghin/models/course/response.ts
var schemaStatus2 = string.transform((value) => value.toUpperCase()).pipe(import_zod12.z.enum(["ACTIVE", "INACTIVE"]));
var schemaCourseCountriesResponse = import_zod12.z.object({
  countries: import_zod12.z.array(schemaCourseCountry.passthrough())
});
var schemaCourseSearchResponse = import_zod12.z.object({
  courses: import_zod12.z.array(schemaCourse.passthrough())
});
var schemaCourseDetailsFacility = import_zod12.z.object({
  FacilityId: number,
  FacilityName: string,
  FacilityNumber: number.nullable(),
  FacilityStatus: string,
  GeoLocationFormattedAddress: schemaGeoAddress,
  GeoLocationLatitude: schemaGeoCoordinate.nullable(),
  GeoLocationLongitude: schemaGeoCoordinate.nullable(),
  GolfAssociationId: number.nullable()
});
var schemaCourseDetailsSeason = import_zod12.z.object({
  IsAllYear: boolean,
  SeasonEndDate: schemaSeasonDate.transform((value) => {
    if (!value) {
      return null;
    }
    const [month, day] = value.split("/");
    return `${month?.toString().padStart(2, "0")}-${day?.toString().padStart(2, "0")}`;
  }),
  SeasonName: schemaSeasonName,
  SeasonStartDate: schemaSeasonDate.transform((value) => {
    if (!value) {
      return null;
    }
    const [month, day] = value.split("/");
    return `${month?.toString().padStart(2, "0")}-${day?.toString().padStart(2, "0")}`;
  })
});
var schemaCourseDetailsTeeSetRatings = import_zod12.z.object({
  BogeyRating: float,
  CourseRating: float,
  RatingType: import_zod12.z.enum(["Front", "Back", "Total"]),
  SlopeRating: float
});
var schemaCourseDetailsTeeSetHole = import_zod12.z.object({
  Allocation: number,
  HoleId: number,
  Length: number,
  Number: number,
  Par: number
});
var schemaCourseDetailsTeeSet = import_zod12.z.object({
  EligibleSides: import_zod12.z.unknown(),
  Gender: import_zod12.z.enum(["Male", "Female", "Mixed"]).nullable(),
  Holes: import_zod12.z.array(schemaCourseDetailsTeeSetHole),
  HolesNumber: number,
  IsShorter: boolean.nullable(),
  LegacyCRPTeeId: number,
  Ratings: import_zod12.z.array(schemaCourseDetailsTeeSetRatings),
  StrokeAllocation: boolean,
  TeeSetRatingId: number,
  TeeSetRatingName: string,
  TotalMeters: number,
  TotalPar: number,
  TotalYardage: number
});
var schemaCourseDetailsResponse = import_zod12.z.object({
  CourseCity: string,
  CourseId: number,
  CourseName: string,
  CourseNumber: number.nullable(),
  CourseState: schemaCourseSearchState,
  CourseStatus: schemaStatus2,
  Facility: schemaCourseDetailsFacility,
  Season: schemaCourseDetailsSeason.nullable(),
  TeeSets: import_zod12.z.array(schemaCourseDetailsTeeSet)
});

// src/client/ghin/models/course/tee-set-rating.ts
var import_zod13 = require("zod");
var schemaStatus3 = string.transform((value) => value.toUpperCase()).pipe(import_zod13.z.enum(["ACTIVE", "INACTIVE"]));
var schemaTeeSetRatingRequest = import_zod13.z.object({
  tee_set_rating_id: number,
  include_altered_tees: boolean.optional().default(true)
});
var schemaTeeSetRatingSeason = import_zod13.z.object({
  SeasonName: schemaSeasonName,
  SeasonStartDate: schemaSeasonDate.transform((value) => {
    if (!value) {
      return null;
    }
    const [month, day] = value.split("/");
    return `${month?.toString().padStart(2, "0")}-${day?.toString().padStart(2, "0")}`;
  }),
  SeasonEndDate: schemaSeasonDate.transform((value) => {
    if (!value) {
      return null;
    }
    const [month, day] = value.split("/");
    return `${month?.toString().padStart(2, "0")}-${day?.toString().padStart(2, "0")}`;
  }),
  IsAllYear: boolean
});
var schemaTeeSetRatingCourse = import_zod13.z.object({
  CourseId: number,
  CourseStatus: schemaStatus3,
  CourseName: string,
  CourseNumber: number.nullable(),
  CourseCity: string,
  CourseState: schemaCourseSearchState
});
var schemaTeeSetRatingFacility = import_zod13.z.object({
  FacilityId: number,
  FacilityStatus: string,
  FacilityName: string,
  FacilityNumber: number.nullable(),
  GolfAssociationId: number.nullable().optional(),
  GeoLocationFormattedAddress: schemaGeoAddress.optional(),
  GeoLocationLatitude: schemaGeoCoordinate.nullable().optional(),
  GeoLocationLongitude: schemaGeoCoordinate.nullable().optional()
});
var schemaTeeSetRatingHole = import_zod13.z.object({
  HoleId: number,
  Number: number,
  Par: number,
  Length: number,
  Allocation: number
});
var schemaTeeSetRatingRating = import_zod13.z.object({
  RatingType: import_zod13.z.enum(["Front", "Back", "Total"]),
  CourseRating: float,
  SlopeRating: float,
  BogeyRating: float
});
var schemaTeeSetRatingResponse = import_zod13.z.object({
  Season: schemaTeeSetRatingSeason.nullable(),
  Course: schemaTeeSetRatingCourse,
  Facility: schemaTeeSetRatingFacility,
  TeeSetRatingId: number,
  TeeSetRatingName: string,
  TeeSetStatus: import_zod13.z.enum(["Active", "Inactive", "Deleted"]).transform((val) => val.toLowerCase()).optional(),
  Gender: import_zod13.z.enum(["Male", "Female", "Mixed"]).nullable(),
  HolesNumber: number,
  TotalPar: number,
  TotalYardage: number,
  TotalMeters: number,
  StrokeAllocation: boolean,
  IsShorter: boolean.nullable(),
  LegacyCRPTeeId: number.nullable(),
  EligibleSides: import_zod13.z.unknown(),
  Holes: import_zod13.z.array(schemaTeeSetRatingHole),
  Ratings: import_zod13.z.array(schemaTeeSetRatingRating)
});

// src/client/ghin/models/facility/facility.ts
var import_zod14 = require("zod");
var schemaStatus4 = string.transform((value) => value.toUpperCase()).pipe(import_zod14.z.enum(["ACTIVE", "INACTIVE"]));
var schemaFacilityCourse = import_zod14.z.object({
  CourseId: number,
  CourseStatus: schemaStatus4,
  CourseName: string,
  NumberOfHoles: number
});
var schemaFacility = import_zod14.z.object({
  Address1: string.nullable().optional(),
  Address2: string.nullable().optional(),
  Associations: import_zod14.z.array(import_zod14.z.unknown()).optional(),
  City: string.nullable(),
  Country: string.nullable(),
  Courses: import_zod14.z.array(schemaFacilityCourse).optional(),
  Email: string.email().nullable().optional(),
  EntCountryCode: number.nullable(),
  EntStateCode: number.nullable(),
  FacilityId: number,
  FacilityName: string,
  FacilityStatus: schemaStatus4,
  GeoLocationLatitude: schemaGeoCoordinate.nullable().optional(),
  GeoLocationLongitude: schemaGeoCoordinate.nullable().optional(),
  State: string.nullable(),
  Telephone: string.nullable().optional(),
  UpdatedOn: shortDate.nullable().optional(),
  Zip: import_zod14.z.string().trim().transform((zip) => zip?.trim() || null).nullable().optional().or(import_zod14.z.array(import_zod14.z.unknown()))
});

// src/client/ghin/models/facility/request.ts
var import_zod15 = require("zod");
var schemaFacilitySearchRequest = import_zod15.z.object({
  country: schemaCourseCountryCode.optional(),
  facility_id: number.optional(),
  name: string.optional(),
  state: schemaCourseSearchState.optional()
}).refine(
  ({ country, state, facility_id, name }) => {
    return Boolean(country || state || facility_id || name);
  },
  {
    message: "At least one search parameter must be provided: country, state, facility_id, or name"
  }
);

// src/client/ghin/models/facility/response.ts
var import_zod16 = require("zod");
var schemaFacilitySearchResponse = import_zod16.z.array(schemaFacility);

// src/client/ghin/models/golfers/search.ts
var import_zod17 = require("zod");
var schemaStatus5 = import_zod17.z.enum(["Active", "Inactive"]);
var schemaGolfersGlobalSearchRequest = import_zod17.z.object({
  country: string.transform((value) => value?.toUpperCase()),
  first_name: string,
  from_ghin: boolean,
  ghin: number,
  last_name: string.optional(),
  order: import_zod17.z.enum(["asc", "desc"]),
  page: number,
  per_page: number.max(100),
  sorting_criteria: import_zod17.z.enum(["country", "full_name", "handicap_index", "state", "status"]),
  state: string.transform((value) => value?.toUpperCase()),
  status: schemaStatus5
}).partial();
var schemaGolfersSearchRequest = import_zod17.z.object({
  page: number,
  per_page: number.max(100),
  golfer_id: number.optional(),
  last_name: string.optional(),
  first_name: emptyStringToNull.optional(),
  state: emptyStringToNull.transform((value) => value?.toUpperCase()).optional(),
  country: string.transform((value) => value?.toUpperCase()).optional(),
  local_number: emptyStringToNull.optional(),
  email: emptyStringToNull.optional(),
  phone_number: emptyStringToNull.optional(),
  association_id: number.optional(),
  club_id: emptyStringToNull.optional(),
  sorting_criteria: import_zod17.z.enum([
    "first_name",
    "last_name",
    "status",
    "id",
    "gender",
    "date_of_birth",
    "handicap_index",
    "status_date",
    "full_name",
    "home_club",
    "last_name_first_name"
  ]).optional(),
  order: import_zod17.z.enum(["asc", "desc"]).optional(),
  status: schemaStatus5.optional(),
  updated_since: emptyStringToNull.optional()
}).partial();
var schemaGolfer = import_zod17.z.object({
  ghin: number,
  first_name: string,
  last_name: string,
  association_id: number,
  association_name: string,
  handicap_index: handicap,
  club_affiliation_id: number,
  club_id: number,
  club_name: emptyStringToNull,
  country: emptyStringToNull,
  entitlement: boolean,
  gender,
  hard_cap: boolean,
  has_digital_profile: boolean,
  hi_display: string,
  hi_value: handicap,
  is_home_club: boolean,
  low_hi_date: date.nullable(),
  low_hi_display: string,
  low_hi_value: handicap,
  low_hi: handicap,
  message_club_authorized: string.nullable(),
  middle_name: emptyStringToNull.nullable().optional(),
  phone_number: emptyStringToNull.nullable().optional(),
  prefix: emptyStringToNull.optional(),
  rev_date: date.nullable(),
  soft_cap: boolean,
  state: emptyStringToNull,
  status: schemaStatus5,
  suffix: emptyStringToNull.optional()
});
var schemaGolfersSearchResponse = import_zod17.z.object({
  golfers: import_zod17.z.array(schemaGolfer)
});

// src/client/ghin/models/handicaps/course-player-handicap.ts
var import_zod18 = require("zod");
var schemaPlayerCourseHandicap = import_zod18.z.object({
  playing_handicap: number,
  playing_handicap_display: string,
  shots_off: number
});
var schemaCoursePercentPlayerHandicap = import_zod18.z.record(string, schemaPlayerCourseHandicap);

// src/client/ghin/models/handicaps/request.ts
var import_zod19 = require("zod");
var schemaTeeSetSide = import_zod19.z.enum(["All 18", "F9", "B9"]);
var schemaGolferCourseHandicapBaseRequest = import_zod19.z.object({
  ghin: number.optional(),
  handicap_index: number.optional(),
  tee_set_id: number,
  tee_set_side: schemaTeeSetSide
});
var schemaGolferCourseHandicapRequest = schemaGolferCourseHandicapBaseRequest.refine(({ ghin, handicap_index }) => {
  if (ghin && handicap_index) {
    throw new Error("Cannot provide both ghin and handicap_index");
  }
  if (!Number.isSafeInteger(ghin) && !Number.isSafeInteger(handicap_index)) {
    throw new Error("Must provide either ghin or handicap_index");
  }
  return true;
});
var schemaCourseHandicapsRequest = import_zod19.z.object({
  golfers: import_zod19.z.array(
    schemaGolferCourseHandicapBaseRequest.omit({
      ghin: true
    }).extend({
      golfer_id: number.optional()
    })
  ),
  source: import_zod19.z.literal(CLIENT_SOURCE).default(CLIENT_SOURCE).optional()
});

// src/client/ghin/models/handicaps/response.ts
var import_zod20 = require("zod");
var schemaGolferHandicapClub = import_zod20.z.object({
  active: boolean,
  association_id: number,
  club_name: string,
  id: number
});
var schemaGolferHandicapResponse = import_zod20.z.object({
  golfer: import_zod20.z.object({
    clubs: import_zod20.z.array(schemaGolferHandicapClub),
    handicap_index: handicap
  })
}).passthrough();
var schemaCoursePlayerHandicapsResponse = import_zod20.z.object({
  100: schemaCoursePercentPlayerHandicap,
  95: schemaCoursePercentPlayerHandicap,
  90: schemaCoursePercentPlayerHandicap,
  85: schemaCoursePercentPlayerHandicap,
  80: schemaCoursePercentPlayerHandicap,
  75: schemaCoursePercentPlayerHandicap,
  70: schemaCoursePercentPlayerHandicap,
  65: schemaCoursePercentPlayerHandicap,
  60: schemaCoursePercentPlayerHandicap,
  55: schemaCoursePercentPlayerHandicap,
  50: schemaCoursePercentPlayerHandicap,
  45: schemaCoursePercentPlayerHandicap,
  40: schemaCoursePercentPlayerHandicap,
  35: schemaCoursePercentPlayerHandicap,
  30: schemaCoursePercentPlayerHandicap,
  25: schemaCoursePercentPlayerHandicap,
  20: schemaCoursePercentPlayerHandicap,
  15: schemaCoursePercentPlayerHandicap,
  10: schemaCoursePercentPlayerHandicap,
  5: schemaCoursePercentPlayerHandicap
});

// src/client/ghin/models/scores/adjustment.ts
var import_zod21 = require("zod");
var schemaScoringAdjustment = import_zod21.z.object({
  display: string,
  type: string,
  value: float
});

// src/client/ghin/models/scores/hole-detail.ts
var import_zod22 = require("zod");
var schemaHoleDetail = import_zod22.z.object({
  adjusted_gross_score: number,
  approach_shot_accuracy: float.nullable().default(0),
  drive_accuracy: float.nullable(),
  fairway_hit: boolean.nullable(),
  gir_flag: import_zod22.z.any(),
  hole_number: number,
  id: number,
  most_likely_score: number.nullable(),
  par: number,
  putts: number.nullable(),
  raw_score: number,
  stroke_allocation: number,
  x_hole: boolean
});

// src/client/ghin/models/scores/request.ts
var import_zod25 = require("zod");

// src/client/ghin/models/scores/score.ts
var import_zod24 = require("zod");

// src/client/ghin/models/scores/statistics.ts
var import_zod23 = require("zod");
var schemaStatistics = import_zod23.z.object({
  birdies_or_better_percent: float,
  bogeys_percent: float,
  double_bogeys_percent: float,
  fairway_hits_percent: float,
  gir_percent: float,
  last_stats_update_date: date,
  last_stats_update_type: emptyStringToNull,
  missed_general_approach_shot_accuracy_percent: float,
  missed_left_approach_shot_accuracy_percent: float,
  missed_left_percent: float,
  missed_long_approach_shot_accuracy_percent: float,
  missed_long_percent: float,
  missed_right_approach_shot_accuracy_percent: float,
  missed_right_percent: float,
  missed_short_approach_shot_accuracy_percent: float,
  missed_short_percent: float,
  one_putt_or_better_percent: float,
  par3s_average: float,
  par4s_average: float,
  par5s_average: float,
  pars_percent: float,
  putts_total: number,
  three_putt_or_worse_percent: float,
  triple_bogeys_or_worse_percent: float,
  two_putt_or_better_percent: float,
  two_putt_percent: float,
  up_and_downs_total: number
});

// src/client/ghin/models/scores/score.ts
var rawScoreTypes = ["A", "C", "E", "H", "N", "P", "T"];
var schemaRawScoreTypes = import_zod24.z.enum(rawScoreTypes);
var scoreTypes = ["AWAY", "COMPETITION", "EXCEPTIONAL", "HOME", "9_HOLE_ROUNDS", "PENALTY", "TOURNAMENT"];
var schemaScoreType = import_zod24.z.enum(scoreTypes);
var scoreTypesMap = {
  A: "AWAY",
  C: "COMPETITION",
  E: "EXCEPTIONAL",
  H: "HOME",
  N: "9_HOLE_ROUNDS",
  P: "PENALTY",
  T: "TOURNAMENT"
};
var schemaScoreTypeWithTransform = schemaRawScoreTypes.transform(
  (value) => scoreTypesMap[value]
);
var scoreStatuses = ["VALIDATED", "UNDER_REVIEW"];
var schemaScoreStatus = import_zod24.z.enum(scoreStatuses);
var rawScoreStatuses = ["Validated", "UnderReview"];
var schemaRawScoreStatus = import_zod24.z.enum(rawScoreStatuses);
var scoreStatusesMap = {
  Validated: "VALIDATED",
  UnderReview: "UNDER_REVIEW"
};
var schemaScoreStatusWithTransform = schemaRawScoreStatus.transform(
  (value) => scoreStatusesMap[value]
);
var schemaScore = import_zod24.z.object({
  adjusted_gross_score: number,
  adjustments: import_zod24.z.array(schemaScoringAdjustment),
  back9_adjusted: number.nullable(),
  back9_course_rating: float.nullable(),
  back9_slope_rating: float.nullable(),
  course_rating: float,
  differential: float,
  edited: boolean,
  exceptional: boolean,
  front9_adjusted: number.nullable(),
  front9_course_rating: float.nullable(),
  front9_slope_rating: float.nullable(),
  gender,
  golfer_id: number,
  hole_details: import_zod24.z.array(schemaHoleDetail),
  id: number,
  is_manual: boolean,
  is_recent: boolean,
  message_club_authorized: string.nullable(),
  net_score_differential: float.nullable(),
  number_of_holes: number,
  number_of_played_holes: number,
  order_number: number,
  parent_id: number.nullable(),
  pcc: float,
  penalty_method: string.nullable(),
  penalty_type: string.nullable(),
  penalty: boolean.optional(),
  played_at: date,
  posted_at: date,
  revision: boolean,
  score_day_order: number,
  score_type_display_full: string,
  score_type_display_short: string,
  score_type: schemaScoreTypeWithTransform,
  season_end_date_at: monthDay,
  season_start_date_at: monthDay,
  slope_rating: float,
  statistics: schemaStatistics.nullable().optional(),
  status: schemaScoreStatusWithTransform,
  unadjusted_differential: float,
  used: boolean
});

// src/client/ghin/models/scores/request.ts
var schemaScoresRequest = import_zod25.z.object({
  course_id: number,
  from_date_played: date,
  limit: number.default(100),
  offset: number.default(0),
  score_types: import_zod25.z.array(import_zod25.z.enum(rawScoreTypes)),
  statuses: import_zod25.z.array(string),
  tee_set_id: number,
  to_date_played: date
}).partial().default({}).optional();

// src/client/ghin/models/scores/response.ts
var import_zod26 = require("zod");
var schemaNumberOrDash = import_zod26.z.union([number, import_zod26.z.literal("-")]).transform((value) => value === "-" ? null : Number(value));
var schemaFloatOrDash = import_zod26.z.union([float, import_zod26.z.literal("-")]).transform((value) => value === "-" ? null : Number.parseFloat(value.toString()));
var schemaScoresResponse = import_zod26.z.object({
  average: schemaFloatOrDash.default(0),
  highest_score: schemaNumberOrDash,
  lowest_score: schemaNumberOrDash,
  scores: import_zod26.z.array(schemaScore),
  total_count: schemaNumberOrDash.default(0)
});

// src/client/ghin/index.ts
var searchParameters = {
  GOLFER_ID: "golfer_id",
  SOURCE: "source"
};
var GhinClient = class {
  httpClient;
  courses;
  facilities;
  golfers;
  handicaps;
  constructor(config) {
    const results = schemaClientConfig.safeParse(config);
    if (!results.success) {
      throw new ConfigurationError(`Invalid GhinClientConfig: ${results.error.message}`);
    }
    this.httpClient = new RequestClient({
      ...results.data,
      cache: results.data.cache ?? new InMemoryCacheClient()
    });
    this.courses = {
      getCountries: this.coursesGetCountries.bind(this),
      getDetails: this.courseGetDetails.bind(this),
      search: this.courseSearch.bind(this),
      getTeeSetRating: this.courseGetTeeSetRating.bind(this)
    };
    this.facilities = {
      search: this.facilitySearch.bind(this)
    };
    this.handicaps = {
      getOne: this.handicapsGetOne.bind(this),
      getCoursePlayerHandicaps: this.handicapsGetCoursePlayerHandicaps.bind(this)
    };
    this.golfers = {
      getOne: this.golfersGetOne.bind(this),
      getScores: this.golfersGetScores.bind(this),
      search: this.golfersSearch.bind(this),
      globalSearch: this.golfersGlobalSearch.bind(this)
    };
  }
  async coursesGetCountries() {
    try {
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "course_countries",
        options,
        schema: schemaCourseCountriesResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value.countries;
    } catch (error) {
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async courseGetDetails(request) {
    try {
      const validRequest = schemaCourseDetailsRequest.parse(request);
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      for (const [key, value] of Object.entries(validRequest)) {
        searchParams.set(key, value.toString());
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "course_details",
        options,
        schema: schemaCourseDetailsResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid course details request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async courseGetTeeSetRating(request) {
    try {
      const validRequest = schemaTeeSetRatingRequest.parse(request);
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      if (validRequest.include_altered_tees !== void 0) {
        searchParams.set("include_altered_tees", validRequest.include_altered_tees.toString());
      }
      const path = `/TeeSetRatings/${validRequest.tee_set_rating_id}.json`;
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetchCustomPath({
        path,
        options,
        schema: schemaTeeSetRatingResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid tee set rating request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async courseSearch(request) {
    try {
      const validRequest = schemaCourseSearchRequest.parse(request);
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      for (const [key, value] of Object.entries(validRequest)) {
        searchParams.set(key, value.toString());
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "course_search",
        options,
        schema: schemaCourseSearchResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value.courses;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid course search request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async facilitySearch(request) {
    try {
      const validRequest = schemaFacilitySearchRequest.parse(request);
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      for (const [key, value] of Object.entries(validRequest)) {
        searchParams.set(key, value.toString());
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "facility_search",
        options,
        schema: schemaFacilitySearchResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid facility search request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async handicapsGetOne(ghin) {
    try {
      const ghinNumber = number.parse(ghin);
      const searchParams = new URLSearchParams([
        ["source", CLIENT_SOURCE],
        ["ghin", ghinNumber.toString()]
      ]);
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "golfer",
        options,
        schema: schemaGolferHandicapResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value.golfer;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid GHIN number: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async handicapsGetCoursePlayerHandicaps(request) {
    try {
      const golfers = import_zod27.z.array(schemaGolferCourseHandicapRequest).parse(request).map(({ ghin, ...golfer }) => ({
        ...golfer,
        [searchParameters.GOLFER_ID]: ghin
      }));
      const searchParams = new URLSearchParams();
      const courseHandicapRequest = {
        golfers,
        source: CLIENT_SOURCE
      };
      const options = {
        body: JSON.stringify(courseHandicapRequest),
        method: "POST",
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "course_handicaps",
        options,
        schema: schemaCoursePlayerHandicapsResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid course handicap request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async golfersSearch(request) {
    try {
      const params = schemaGolfersSearchRequest.parse(request);
      const searchParams = new URLSearchParams();
      const searchDefaults = {
        page: 1,
        per_page: 25,
        sorting_criteria: "last_name_first_name",
        status: "Active",
        order: "asc"
      };
      for (const [key, value] of Object.entries(searchDefaults)) {
        searchParams.set(key, value.toString());
      }
      for (const [key, value] of Object.entries(params)) {
        searchParams.set(key, value?.toString() ?? "");
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "golfers_search",
        schema: schemaGolfersSearchResponse,
        options
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value.golfers;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid golfer search request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async golfersGlobalSearch(request) {
    try {
      const { ghin } = schemaGolfersGlobalSearchRequest.parse(request);
      const searchParams = new URLSearchParams([["source", CLIENT_SOURCE]]);
      const searchDefaults = {
        from_ghin: true,
        per_page: 25,
        sorting_criteria: "full_name",
        order: "asc",
        page: 1
      };
      for (const [key, value] of Object.entries(searchDefaults)) {
        searchParams.set(key, value.toString());
      }
      if (ghin) {
        searchParams.set(searchParameters.GOLFER_ID, ghin.toString());
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "golfers_global_search",
        schema: schemaGolfersSearchResponse,
        options
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value.golfers;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid golfer search request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async golfersGetOne(ghinNumber) {
    try {
      const ghin = number.parse(ghinNumber);
      const results = await this.golfersGlobalSearch({
        ghin,
        status: "Active"
      });
      return results.find((golfer) => golfer.status === "Active");
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid GHIN number: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
  async golfersGetScores(ghinNumber, request) {
    try {
      const validRequest = schemaScoresRequest.parse(request) ?? {};
      const ghin = number.parse(ghinNumber);
      const searchParams = new URLSearchParams([
        [searchParameters.GOLFER_ID, ghin.toString()],
        ["source", CLIENT_SOURCE]
      ]);
      for (const [key, value] of Object.entries(validRequest)) {
        if (value === null) {
          continue;
        }
        if (Array.isArray(value)) {
          for (const v of value) {
            searchParams.append(key, v.toString());
          }
          continue;
        }
        if (typeof value === "object" && value instanceof Date) {
          searchParams.set(key, value.toISOString().split("T")[0]);
          continue;
        }
        searchParams.set(key, value.toString());
      }
      const options = {
        searchParams
      };
      const result = await this.httpClient.fetch({
        entity: "scores",
        options,
        schema: schemaScoresResponse
      });
      if (result.isErr()) {
        throw result.error;
      }
      return result.value;
    } catch (error) {
      if (error instanceof import_zod27.z.ZodError) {
        throw new ValidationError(`Invalid scores request: ${error.message}`);
      }
      throw error instanceof Error ? error : new Error(String(error));
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthenticationError,
  CacheError,
  ConfigurationError,
  ErrorCodes,
  GhinClient,
  GhinError,
  NetworkError,
  RateLimitError,
  ValidationError,
  boolean,
  date,
  emptyStringToNull,
  float,
  gender,
  handicap,
  monthDay,
  number,
  schemaCacheClient,
  schemaClientConfig,
  schemaCourseCountriesResponse,
  schemaCourseCountry,
  schemaCourseCountryCode,
  schemaCourseDetailsRequest,
  schemaCourseDetailsResponse,
  schemaCourseHandicapsRequest,
  schemaCoursePercentPlayerHandicap,
  schemaCoursePlayerHandicapsResponse,
  schemaCourseSearchRequest,
  schemaCourseSearchResponse,
  schemaFacility,
  schemaFacilityCourse,
  schemaFacilitySearchRequest,
  schemaFacilitySearchResponse,
  schemaGolfer,
  schemaGolferCourseHandicapRequest,
  schemaGolferHandicapResponse,
  schemaGolfersGlobalSearchRequest,
  schemaGolfersSearchRequest,
  schemaGolfersSearchResponse,
  schemaHoleDetail,
  schemaPlayerCourseHandicap,
  schemaScoresRequest,
  schemaScoresResponse,
  schemaScoringAdjustment,
  schemaStatistics,
  schemaTeeSetRatingRequest,
  schemaTeeSetRatingResponse,
  schemaTeeSetSide,
  shortDate,
  string
});
